tasks/main.yml

```diff
@@ -1,39 +1,129 @@
     state: present
+  tags: packages
+  notify: __vpn_handler_enable_start_vpn
+
+- name: Controller tasks
+  block:
+    - name: Enforce default auth method as needed
+      set_fact:
+        vpn_connections: |
+          {% for tunnel in vpn_connections %}
+          {%   set _ = tunnel.__setitem__(
+            "auth_method", tunnel.auth_method | d(vpn_auth_method)
+            ) %}
+          {% endfor %}
+          {{ vpn_connections }}
+
+    - name: Make sure that the hosts list is not empty
+      vars:
+        failure: >-
+          {%- for tunnel in vpn_connections %}
+          {%   if not tunnel.hosts -%}True{%- endif %}
+          {% endfor %}
+      fail:
+        msg: list of hosts is empty for one or more tunnels
+      when: '"True" in failure'
+
+    - name: Ensure cert_names are populated when auth_method is cert
+      vars:
+        failure: >-
+          {%- for tunnel in vpn_connections %}
+          {%   if tunnel.auth_method == 'cert' %}
+          {%     if tunnel.hosts is mapping %}
+          {%       for key, value in tunnel.hosts.items() %}
+          {%-        if value is none or 'cert_name' not in value or value.cert_name is none or not value.cert_name | trim | length -%}
+          True
+          {%-        endif -%}
+          {%       endfor %}
+          {%     else -%}True{%-endif %}
+          {%   endif %}
+          {% endfor %}
+      fail:
+        msg: cert_name is missing or empty for one or more hosts in a tunnel
+      when: '"True" in failure'
+
+    - name: generate psks on first host
+      run_once: true
+      set_fact:
+        __vpn_psks: |
+          {% set __vpn_psks = {} %}
+          {% for tunnel in vpn_connections | d([]) %}
+          {%   set __vpn_idx = loop.index0 %}
+          {%   if tunnel.auth_method == 'psk' %}
+          {%     set _ = __vpn_psks.__setitem__(__vpn_idx, {}) %}
+          {%     for host1, host2 in tunnel.hosts.keys() | combinations(2) %}
+          {%       if not host1 in __vpn_psks[__vpn_idx] %}
+          {%         set _ = __vpn_psks[__vpn_idx].__setitem__(host1, {}) %}
+          {%       endif %}
+          {%       if not host2 in __vpn_psks[__vpn_idx] %}
+          {%         set _ = __vpn_psks[__vpn_idx].__setitem__(host2, {}) %}
+          {%       endif %}
+          {%       set psk = lookup('lines', 'openssl rand -base64 48') %}
+          {%       set val = {'pre_shared_key':psk} %}
+          {%       set _ = __vpn_psks[__vpn_idx][host1].__setitem__(host2, val) %}
+          {%       set _ = __vpn_psks[__vpn_idx][host2].__setitem__(host1, val) %}
+          {%     endfor %}
+          {%   endif %}
+          {% endfor %}
+          {{ __vpn_psks }}
+
+    - name: set psks for all hosts
+      set_fact:
+        __vpn_psks: "{{ hostvars[hostvars.keys() | first]['__vpn_psks'] }}"
+      run_once: true
+
+    - name: show it
+      debug:
+        msg: __vpn_psks {{ __vpn_psks | to_nice_json }}
+
+  delegate_to: localhost
+  run_once: true
```

Comment:
if we have a `delegate_to: localhost` and `run_once: true` on the block, do we also need to set `run_once: true` on individual tasks inside the block?

Reply:
No, it applies to all of the tasks inside the block as well. I had accidentally set `run_once: true` for one of the tasks, but I will remove that since it's redundant